feat: Add way to add clipboard providers

64-bitman opened the pull request:
Should fix #12419. Since we already have the `clipmethod` option now, so we use that. The user declares a clipboard provider in the `v:clipproviders` dictionary, where each key is the name of the provider and the value is another dictionary containing some callbacks. The name of each provider can then be used in the `clipmethod` option.

brianhuster commented:
Just in case Nvim compatibility is considered, this is the document of `g:clipboard`, which I think is equivalent with `'clipfunc'` that you propose. https://neovim.io/doc/user/provider.html#g:clipboard

64-bitman commented:
> Just in case Nvim compatibility is considered, this is the document of `g:clipboard` in Nvim, which I think is equivalent with `'clipfunc'` that you propose. https://neovim.io/doc/user/provider.html#g:clipboard

Looking at it, maybe there could be a `v:clipboard` vim var that is a list of dictionaries (providers), each with a paste, copy, and a function that checks if the provider is available + a name for the provider. Then those names of the providers can be used in `clipmethod`. Don't know if that is just overcomplicated but sounds straightforward enough

64-bitman commented:
> > and a function that checks if the provider is available
> 
> If you do so, please don't query for executable existence as it may slowdown vim startup on windows. It is quite slow there. I remember this was the issue with neovim back in a day I tried to use it on my windows box.

Thanks, I will take that into consideration

64-bitman committed:
add clipboard provider feature

h-east commented on src/testdir/test_clipboard_provider.vim:
```diff
@@ -0,0 +1,81 @@
+  set clipmethod=test
+
+  if has('unnamedplus')
+      call assert_equal("reg: +", getreg('+')) 
```
CI error has occurred.

```
From test_codestyle.vim:
Found errors in Test_test_files():
command line..script /home/runner/work/vim/vim/src/testdir/runtest.vim[636]..function RunTheTest[63]..Test_test_files[33]..<SNR>8_ReportError line 2: testdir/test_clipboard_provider.vim line 74: trailing white space
make[2]: *** [Makefile:187: opt_test.vim] Error 1
```

```suggestion
      call assert_equal("reg: +", getreg('+'))
```

64-bitman committed:
use different design

chrisbra commented:
can you please regenerate configure script on latest HEAD?

64-bitman commented:
Turning this into a draft again for now because I've decided to go a different direction with the design

64-bitman committed:
copy and paste callbacks working

64-bitman committed:
refactor more code

- make getting the callback have its own function
- simplify error handling

64-bitman committed:
fix bugs

now works properly

64-bitman committed:
make paste callback use list of strings not a string

64-bitman committed:
remove error message

64-bitman committed:
update docs, refactor

- explicitly reset the clipboard when availability of one of the
  registers changes.
- add "implicit" and "explicit" args to paste function
- add register name argument to both functions
- start creating tests

64-bitman committed:
small update to docs

64-bitman committed:
add tests

64-bitman committed:
Fix bugs

64-bitman committed:
revert unknown changes

64-bitman committed:
make "other" clipmethod only work if not on unix

ychin commented:
Also, what's the point of the "gui" clip method? In what situation does it serve a purpose than the other methods? It basically calls the same functionality as the other methods. Is it just to make it so GVim will automatically swap to GUI when starting up, but otherwise the user may not want to initialize Wayland/X11? I don't think the documentation is describing the intent enough. If that's the case shouldn't the default have "gui,wayland,x11,other" rather than "wayland,x11,gui,other"?

I'm also struggling to understand this piece of code:

```c
#if defined(FEAT_GUI) && defined(FEAT_WAYLAND)
    if (method == CLIPMETHOD_GUI)
	// We only interact with Wayland for the clipboard, we can just deinit
	// everything.
	wayland_uninit_connection();
#endif
```

If the clip method is "GUI" doesn't that mean the GUI is indeed initialized? Why are we de-initializing Wayland here?

ychin commented:
> I plan on refactoring the code to address these issues when I have time, see this https://github.com/vim/vim/issues/18600#issuecomment-3444260510. Essentially I will remove the "gui" and "other" values from clipmethod, as they were made for the clipboard feature to work via the clipmethod option. I'll move the clipboard provider code to not be under the clipboard feature to get rid of all these confusing ifdefs.

I commented on the linked issue, but just duplicating it here. I think what you said sounds good. The current "gui" / "other" are quite confusing and they seem to be from a consequence of the option being originally from just supporting Wayland clipboards (for example, "other" actually means the actual platform native method). 

That said, if this isn't fixed in less than 1 day I think we should just revert it. Things should not remain broken on master branch. And it gives you more time to properly fix and address the design rather than having to hack around it under time pressure.

Shane-XB-Qian commented:
> That said, if this isn't fixed in less than 1 day I think we should just revert it. Things should not remain broken. And it gives you more time to properly fix it rather than having to hack around it under time pressure.

Looking this and think what i said in #18583 and #18498,
so am i a `!operator`?

or @girishji you are a "mad" `joker`?
or @nickspoons do you plan to "kick" something?
or @chrisbra how about this is the "last" to warn me sth?

dear jokers, very enjoying~
***@***.*** how is this feeling? offensive or condescending?)
***@***.*** what is this feeling to mix tech with political/manner like playing in your stackoverflow or "reddit"?)

p.s: @64-bitman @ychin sry, i took this as as example.

-- 
shane.xb.qian
